pkg/verifier/separated_blocks.go

```diff
@@ -0,0 +1,162 @@
+package verifier
+
+import (
+	"context"
+	"time"
+
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/block"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb"
+)
+
+const SeparatedBlocksID = "separated_blocks"
+
+// TODO(bplotka): Move that to flag?
+var MinBlockSize = 2 * time.Hour
+
+// SeparatedBlocks verifies and fixes (if repair is true) the issue with block being missed and assumed that no block is
+// there for given time range.
+// Example:
+//  - a block with range 0 - 100
+//  - another ovelapped block with range 20-50
+//  - bigger block is missing exactly these sources from smaller block to fill its range.
+// Various bugs could introduce these blocks, including broken compaction for Prometheus 2.2.0.
+func SeparatedBlocks(ctx context.Context, logger log.Logger, bkt objstore.Bucket, backupBkt objstore.Bucket, repair bool) error {
```

Comment:
Mh, can you elaborate a bit more what motivated this one. It seems like the starting position the example describes is actually valid, i.e. not an issue.

If the compactor does not pick that input up for compaction, that's probably just: https://github.com/prometheus/tsdb/issues/90

Though it can of course involve risk if the input blocks aren't truly disjoint in their actual data.

Reply:
The example states the broken state (overlapping), why you say it is not an issue?

Thanks for referencing this issue. It all makes sense, so maybe we want to add something like that for TSDB? I mean at least explicit `VerticalCompact` for now not used internally yet. To compact overlapped blocks.

Wonder however if compactor should do that automatically. For now any overlap case is invalid state. It can be introduced only by bug. It should not happen. Of course we could make it differently -> allow or don't care about overlaps and just "heal" them by something like vertical compact, true, but not sure if that's the right path. Not for now for sure.

For now I would really merge this for now here.